### PR TITLE
New Config: worker_enable_prefetch_count_reduction

### PR DESCRIPTION
**Pull request by Nusnus**

### CODE REVIEW

**Review by auvipy — COMMENTED** [review]
should we consider it as a missing feature for v5.3.x or next v5.4.x?

**Comment by Nusnus** [comment]
> should we consider it as a missing feature for v5.3.x or next v5.4.x?

Definitely 5.3.x.

This PR doesn't break backwards compatibility 💪

File: celery/app/defaults.py [hunk 1]
```diff
@@ -325,6 +325,7 @@ def __repr__(self):
         pool_restarts=Option(False, type='bool'),
         proc_alive_timeout=Option(4.0, type='float'),
         prefetch_multiplier=Option(4, type='int'),
+        enable_prefetch_count=Option(True, type='bool'),
```
Comment (thedrow) [inline celery/app/defaults.py:328]: This is not a meaningful name for the feature flag.
Prefetching is always enabled.
What this feature flag controls is the reduction of the prefetch count when the connection restarts.

File: celery/worker/consumer/consumer.py [hunk 2]
```diff
@@ -696,7 +697,10 @@ def on_task_received(message):
 
     def _restore_prefetch_count_after_connection_restart(self, p, *args):
         with self.qos._mutex:
-            if self._maximum_prefetch_restored:
+            if any([
+                not self.app.conf.worker_enable_prefetch_count,
+                self._maximum_prefetch_restored,
+            ]):
```
Comment (thedrow) [inline celery/worker/consumer/consumer.py:703]:
```suggestion
            if any((
                not self.app.conf.worker_enable_prefetch_count,
                self._maximum_prefetch_restored,
            )):
```
Use a tuple as this is more efficient to initialize.

File: docs/userguide/configuration.rst [hunk 3]
```diff
@@ -2969,6 +2970,48 @@ For more on prefetching, read :ref:`optimizing-prefetch-limit`
 
     Tasks with ETA/countdown aren't affected by prefetch limits.
 
+.. setting:: worker_enable_prefetch_count
+
+``worker_enable_prefetch_count``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. versionadded:: 5.3
```
Comment (thedrow) [inline docs/userguide/configuration.rst:2978]: This should be in 5.4 as this is a new feature.

**Review by thedrow — CHANGES_REQUESTED** [review]

File: celery/app/defaults.py (hunk 1 above)
Comment (Nusnus) [reply celery/app/defaults.py:328]: Can you offer a better name @thedrow ?

File: docs/userguide/configuration.rst (hunk 3 above)
Comment (Nusnus) [reply docs/userguide/configuration.rst:2978]: Why?
This flag does not break compatibility, and defacto leaves the same behavior due to defaults.

Comment (thedrow) [reply docs/userguide/configuration.rst:2978]: Patch releases are reserved for bug fixes. There is no bug to fix here.
Minor releases are for new features that don't break backward compatibility.

File: celery/app/defaults.py (hunk 1 above)
Comment (thedrow) [reply celery/app/defaults.py:328]: Maybe `enable_prefetch_count_reduction_on_connection_restart`?

**Comment by Nusnus** [comment]
> > should we consider it as a missing feature for v5.3.x or next v5.4.x?
> 
> Definitely 5.3.x.
> 
> This PR doesn't break backwards compatibility 💪

Revert hehe 

@auvipy I'll fix the PR per the comments from @thedrow but I will only merge after you finish with the v5.3.5 release.

File: docs/userguide/configuration.rst (hunk 3 above)
Comment (Nusnus) [reply docs/userguide/configuration.rst]: Fix pushed

File: celery/app/defaults.py (hunk 1 above)
Comment (Nusnus) [reply celery/app/defaults.py]: `enable_prefetch_count_reduction_on_connection_restart` was too long. Went for:
`enable_prefetch_count_reduction`

**Comment by Nusnus** [comment]
@thedrow please review again (but don't merge).

**Review by thedrow — APPROVED** [review]

**Comment by Nusnus** [comment]
Rebased on `main`

**Comment by Nusnus** [comment]
> terminate called after throwing an instance of 'std::bad_cast'
    what():  std::bad_cast
  Fatal Python error: Aborted

CI failures are due to internal errors.

**Comment by Nusnus** [comment]
Rebased on `main`

**Comment by Nusnus** [comment]
Should be merged after 5.3.5 is released

**Comment by Nusnus** [comment]
@auvipy This PR should be merged for v5.4.
I was waiting for v5.3.5 to be released before we can start merging 5.4 stuff.

Can we move forward, or is there a v5.3.6 on the way? - I saw some comment somewhere, so I am confirming first.

P.S
Just rebased on `main` and the CI is all green so far as expected

**Comment by auvipy** [comment]
Wait let me check and confirm few hours please

**Comment by Nusnus** [comment]
Docker build failed on timeout.
I'll increase the job timeout in another PR, but this shouldn't block merge.

**Comment by Nusnus** [comment]
Rebased on Celery v5.3.6 (`main`).
If all CI passes successfully, I'll merge.

This will be the first Celery v5.4 merged PR.

**Pull request merged**